app/controllers/users.rb

```diff
@@ -0,0 +1,40 @@
+class MakersBnb < Sinatra::Base
+  delete '/' do 
+    session.clear
+    redirect ('/')
+  end 
+
+  post '/user/new' do
+    user = User.create(first_name: params[:first_name],
+                       last_name: params[:last_name],
+                       email: params[:email],
+                       password: params[:password],
+                       password_confirmation: params[:password_confirmation])
+    if user.saved? 
+      session[:user] = user.id
+      send_mail
+      redirect('/spaces')
+    else 
+      flash.next[:errors] = user.errors.full_messages
+      redirect('/')
+    end
+  end
+
+  get '/user/sign_in' do 
+    erb :'users/sign_in'
+  end
+
+  post '/user/sign_in' do 
+    @user = User.authenticate(params[:email],params[:password])
+    if @user == :nonexistent_user
+      flash.next[:errors] = ["User does not exist"]
+      redirect('/user/sign_in')
+    elsif @user == :incorrect_password 
```

Comment:
Nice idea, I see what you're doing here. I suggest you could handle this more elegantly than in an `if`/`else` block. How about attaching the failure reason to a `@errors` attribute on `User`? Push it down the stack. The `@errors` attribute could look something like this (called via attr_reader): `user.errors # => [{ message: "User does not exist" }]`

Then you'd handle it here like this:

```ruby
post '/user/sign_in' do
  @user = User.authenticate(params)
  return redirect('/users/sign_in'), notice: user.errors if user.errors
  sign_in_and_redirect(user) # legit use of a helper that sets the session and redirects
end
```